src/WireBST.java

```diff
@@ -109,29 +109,242 @@ private void insertToRight(WireBSTNode currentNode , WireBSTNode newNode ){
+    private void updateMedianRemove(WireBSTNode node){
+
+        if (this.getNodeCounter() % 2 == 0) {
+
+            if (node.getStudentNum() <= this.getMedianWireBST().getStudentNum())
+                this.setMedian(getSuccessor(this.getMedianWireBST()));
+
+        }
+
+        else  {
+
+            if (node.getStudentNum() >= this.getMedianWireBST().getStudentNum())
+            this.setMedian(getPredecessor(this.getMedianWireBST()));
+
+        }
+
+        this._nodeCounter --;
+
+    }
+
+    /**
+     * removing a node with only right son
+     * @param node the node we wish to remove
+     */
+    private void removeNodeWithOnlyRightSon(WireBSTNode node){
+
+        if (node.getParent() == null){
+
+            updateMedianRemove(node);
+            node.getRight().setParent(node.getParent());
+            this.setHead(node.getRight());
+            getSuccessor(node).setLeft(node.getLeft());
+
+        }
+
+        else if  (node.equal(node.getParent().getRight())){
+
+            updateMedianRemove(node);
+            node.getRight().setParent(node.getParent());
+            node.getParent().setRight(node.getRight());
+            getSuccessor(node).setLeft(node.getParent());
+
+        }
+
+        else {
+
+            updateMedianRemove(node);
+            node.getRight().setParent(node.getParent());
+            node.getParent().setLeft(node.getRight());
+            getSuccessor(node).setLeft(getPredecessor(node));
+
+        }
+    }
+
+
+
+    /**
+    * removing a node with only left son
+    * @param node the node we wish to remove
+    */
+    private void removeNodeWithOnlyLeftSon(WireBSTNode node){
+
+        if (node.getParent() == null){
+
+            updateMedianRemove(node);
+            this.setHead(node.getLeft());
+            node.getLeft().setParent(node.getParent());
+            getPredecessor(node).setRight(node.getRight());
+
+        }
+
+        else if (node.equal(node.getParent().getRight())){
+
+            updateMedianRemove(node);
+            node.getLeft().setParent(node.getParent());
+            node.getParent().setRight(node.getLeft());
+            getPredecessor(node).setRight(getSuccessor(node));
+
         }
 
+        else {
 
+            updateMedianRemove(node);
+            node.getParent().setLeft(node.getLeft());
+            node.getLeft().setParent(node.getParent());
+            getPredecessor(node).setRight(getSuccessor(node));
 
+        }
 
     }
 
-    // remove of the node WireBST
-    public void removeWireBST(int studetNum){
+    /**
+     * remove node who has two sons
+     * @param node with two sons
+     */
+    private void removeNodeWithTwoSons(WireBSTNode node){
+
+        WireBSTNode temp = getSuccessor(node);
+
+        if (node.getParent() == null ){
+
+            this.setHead(temp);
+            temp.setParent(node.getParent());
+            temp.setRight(node.getRight());
+            node.getRight().setParent(temp);
+            temp.setLeft(node.getLeft());
```

Comment:
temp.getParent().setLeft(temp.getRight());

Reply:
// update the parent of the temp, to point on the right son of temp (temp cant have a left son, cause he is the successor of the nodeToRemove)